dhcpv4/client.go

```diff
@@ -42,12 +46,20 @@ func NewClient() *Client {
 	}
 }
 
-// MakeRawBroadcastPacket converts payload (a serialized DHCPv4 packet) into a
-// raw packet suitable for UDP broadcast.
+// MakeRawBroadcastPacket leverages MakeRawPacket to create a raw packet suitable
+// for UDP broadcast.
 func MakeRawBroadcastPacket(payload []byte) ([]byte, error) {
+	serverAddr := &net.UDPAddr{IP: net.IPv4bcast, Port: ServerPort}
+	clientAddr := &net.UDPAddr{IP: net.IPv4zero, Port: ClientPort}
+	return MakeRawPacket(payload, serverAddr, clientAddr)
+}
+
+// MakeRawPacket converts a payload (a serialized DHCPv4 packet) into a
+// raw packet for the specified serverAddr from the specified clientAddr.
+func MakeRawPacket(payload []byte, serverAddr, clientAddr *net.UDPAddr) ([]byte, error) {
 	udp := make([]byte, 8)
-	binary.BigEndian.PutUint16(udp[:2], ClientPort)
-	binary.BigEndian.PutUint16(udp[2:4], ServerPort)
+	binary.BigEndian.PutUint16(udp[:2], uint16(clientAddr.Port))
```

Comment:
this line and the one below will cause a panic if `clientAddr` or `serverAddr` are `nil`.

At the same time I'm not convinced we should use a pointer to `net.UDPAddr`, it feels redundant, since client and server addresses are required fields.

Another option could be to use `net.Addr` here instead (so the whole higher-level code always uses `net.Addr` objects, which is easier to follow) and let the bottom functions, like this one, deal with the details. Being an interface, it would still require a check against `nil`, and the type assertion should be moved here.